README
NAME
    Text::TabularDisplay - Display text in formatted table output

SYNOPSIS
        use Text::TabularDisplay;

        my $table = Text::TabularDisplay->new(@columns);
        $table->add(@row)
            while (@row = $sth->fetchrow);
        print $table->render;

        +----+-----------+
        | id | name      |
        +----+-----------+
        | 1  | Tom       |
        | 2  | Dick      |
        | 3  | Harry     |
        +----+-----------+

DESCRIPTION
    Text::TabularDisplay simplifies displaying textual data in a table. The
    output is identical to the columnar display of query results in the
    mysql text monitor. For example, this data:

        1, "Tom Jones", "(666) 555-1212"
        2, "Barnaby Jones", "(666) 555-1213"
        3, "Bridget Jones", "(666) 555-1214"

    Used like so:

        my $t = Text::TabularDisplay->new(qw(id name phone));
        $t->add(1, "Tom Jones", "(666) 555-1212");
        $t->add(2, "Barnaby Jones", "(666) 555-1213");
        $t->add(3, "Bridget Jones", "(666) 555-1214");
        print $t->render;

    Produces:

        +----+---------------+----------------+
        | id | name          | phone          |
        +----+---------------+----------------+
        | 1  | Tom Jones     | (666) 555-1212 |
        | 2  | Barnaby Jones | (666) 555-1213 |
        | 3  | Bridget Jones | (666) 555-1214 |
        +----+---------------+----------------+

METHODS
    Text::TabularDisplay has four primary methods: new(), columns(), add(),
    and render(). new() creates a new Text::TabularDisplay instance;
    columns() sets the column headers in the output table; add() adds data
    to the instance; and render() returns a formatted string representation
    of the instance.

    There are also a few auxilliary convenience methods: clone(), items(),
    reset(), populate(), and paginate().

    new A Text::TabularDisplay instance can be created with column names
        passed as constructor args, so these two calls produce similar
        objects:

            my $t1 = Text::TabularDisplay->new;
            $t1->columns(qw< one two >);

            my $t2 = Text::TabularDisplay->new(qw< one two >);

        Calling new() on a Text::TabularDisplay instance returns a clone of
        the object. See the clone entry in the Text::TabularDisplay manpage.

    columns
        Gets or sets the column names for an instance. This method is called
        automatically by the constructor with any parameters that are passed
        to the constructor (if any are passed).

        When called in scalar context, columns() returns the *number of
        columns in the instance*, rather than the columns themselves. In
        list context, copies of the columns names are returned; the names of
        the columns cannot be modified this way.

    add Takes a list of items and appends it to the list of items to be
        displayed. add() can also take a reference to an array, so that
        large arrays don't need to be copied.

        As elements are processed, add() maintains the width of each column
        so that the resulting table has the correct dimensions.

        add() returns $self, so that calls to add() can be chained:

            $t->add(@one)->add(@two)->add(@three);

    render
        render() does most of the actual work. It returns a string
        containing the data added via add(), formatted as a table, with a
        header containing the column names.

        render() does not change the state of the object; it can be called
        multiple times, with identical output (including identical running
        time: the output of render is not cached).

        If there are no columns defined, then the output table does not
        contains a row of column names. Compare these two sequences:

            my $t = Text::TabularDisplay->new;
            $t->add(qw< 1 2 3 4 >);
            $t->add(qw< 5 6 7 8 >);
            print $t->render;

            $t->columns(qw< one two three four >);
            print $t->render;

            # Example 1 output
            +---+---+---+---+
            | 1 | 2 | 3 | 4 |
            | 5 | 6 | 7 | 8 |
            +---+---+---+---+

            # Example 2 output
            +-----+-----+-------+------+
            | one | two | three | four |
            +-----+-----+-------+------+
            | 1   | 2   | 3     | 4    |
            | 5   | 6   | 7     | 8    |
            +-----+-----+-------+------+

        render() takes optional $start and $end arguments; these indicate
        the start and end *indexes* for the data to be rendered. This can be
        used for paging and the like:

            $t->add(1, 2, 3)->add(4, 5, 6)->add(7, 8, 9)->add(10, 11, 12);
            print $t->render(0, 1), "\n";
            print $t->render(2, 3), "\n";

        Produces:

            +-------+--------+-------+
            | First | Second | Third |
            +-------+--------+-------+
            | 1     | 2      | 3     |
            | 4     | 5      | 6     |
            +-------+--------+-------+

            +-------+--------+-------+
            | First | Second | Third |
            +-------+--------+-------+
            | 7     | 8      | 9     |
            | 10    | 11     | 12    |
            +-------+--------+-------+

        As an aside, note the chaining of calls to add().

        The elements in the table are padded such that there is the same
        number of items in each row, including the header. Thus:

            $t->columns(qw< One Two >);
            print $t->render;

            +-----+-----+----+
            | One | Two |    |
            +-----+-----+----+
            | 1   | 2   | 3  |
            | 4   | 5   | 6  |
            | 7   | 8   | 9  |
            | 10  | 11  | 12 |
            +-----+-----+----+

        And:

            $t->columns(qw< One Two Three Four>);
            print $t->render;

            +-----+-----+-------+------+
            | One | Two | Three | Four |
            +-----+-----+-------+------+
            | 1   | 2   | 3     |      |
            | 4   | 5   | 6     |      |
            | 7   | 8   | 9     |      |
            | 10  | 11  | 12    |      |
            +-----+-----+-------+------+

OTHER METHODS
    clone()
        The clone() method returns an identical copy of a
        Text::TabularDisplay instance, completely separate from the cloned
        instance.

    items()
        The items() method returns the number of elements currently stored
        in the data structure:

            printf "There are %d elements in \$t.\n", $t->items;

    reset()
        Reset deletes the data from the instance, including columns. If
        passed arguments, it passes them to columns(), just like new().

    populate()
        populate() as a special case of add(); populate() expects a
        reference to an array of references to arrays, such as returned by
        DBI's selectall_arrayref method:

            $sql = "SELECT " . join(", ", @c) . " FROM mytable";
            $t->columns(@c);
            $t->populate($dbh->selectall_arrayref($sql));

        This is for convenience only; the implementation maps this to
        multiple calls to add().

NOTES / ISSUES
    Text::TabularDisplay assumes it is handling strings, and does stringy
    things with the data, like legnth() and sprintf(). Non-character data
    can be passed in, of course, but will be treated as strings; this may
    have ramifications for objects that implement overloading.

    Newlines contained within the data ruin formatting. This is consistent
    with the software that this module is emulating (the mysql command line
    client), but only because this was easier to implement.

    Currently, this:

        my $t = Text::TabularDisplay->new("id", "text");
        $t->add(1, "This is my text.\nThere are many like it, but this one is mine.");
        print $t->render;

    renders as this:

        +----+----------------------------------------------------------------+
        | id | text
        |
        +----+----------------------------------------------------------------+
        | 1  | This is my text.
        There are many like it, but this one is mine. |
        +----+----------------------------------------------------------------+

    I would like it to render like so:

        +----+----------------------------------------------------------------+
        | id | text                                                           |
        +----+----------------------------------------------------------------+
        | 1  | This is my text.                                               |
        |    | There are many like it, but this one is mine.                  |
        +----+----------------------------------------------------------------+

    Patches welcome, of course. ;)

    The biggest issue, though, is that this module duplicates a some of the
    functionality of Data::ShowTable. Of course, Data::ShowTable is a large,
    complex monolithic tool that does a lot of things, while
    Text::TabularDisplay is small and fast.

VERSION
    $Id: README,v 1.1.1.1 2005/10/04 16:39:58 dlc Exp $

AUTHOR
    darren chamberlain <darren@cpan.org>



Changes
======================================================================
                     --- Text-TabularDisplay ---
======================================================================

2004-10-20 16:28  dlc

	* TabularDisplay (1.20): Applied a mutate agglomeration of a pair of
	patches from Prakash Kailasa and Paul Cameron to address newline folding
	within values, slightly modified according to suggestions from Slaven
	Rezic.

2004-09-21 09:14  dlc

	* TabularDisplay.pm (1.19): Applied a patch from Eric Cholet to
	silence warnings (<https://rt.cpan.org/Ticket/Display.html?id=7693>)

2002-10-28 15:14  dlc

	* README (1.6): Version updates.

2002-10-25 00:10  dlc

	* README (1.5): Updated.

2002-10-25 00:10  dlc

	* Changes (1.11): Automatically generated by cvs2cl.pl for
	dlc@tumbleweed

2002-10-25 00:10  dlc

	* Makefile.PL (1.5): Using DISTVNAME and SUFFIX.

2002-10-25 00:10  dlc

	* MANIFEST (1.7): Added another example: tdisplay

2002-10-25 00:08  dlc

	* examples/mysql.pl (1.3): Added optional addition of
	MySQL::Config.	Changed username references to user, so that the
	defaults in my.cnf could be used.

2002-10-24 23:56  dlc

	* examples/tdisplay (1.2): Hey, this is actually kind of useful. 
	Should this be one of the examples, or should it be an actual part
	of the package?

2002-10-24 23:29  dlc

	* examples/tdisplay (1.1): Command-line script to display stuff as
	tables.

2002-10-24 18:26  dlc

	* t/13.t (1.2): Updated test.

2002-10-24 18:26  dlc

	* TabularDisplay.pm (1.18): Added _SIZE.  Does The Right Thing for
	cases where the number of elements in the header is changed after
	data is added.	See t/13.t for more details, and an example of what
	this actually means.

2002-10-24 17:48  dlc

	* README (1.4): Up-to-date README.

2002-10-24 17:42  dlc

	* Changes (1.10): Automatically generated by cvs2cl.pl for
	dlc@tumbleweed

2002-10-24 17:40  dlc

	* t/14.t (1.1): Test for populate().

2002-10-24 17:40  dlc

	* examples/dump-passwd.pl (1.2): Removed alignment characters (no
	longer applicable), and simplified the program a lot.

2002-10-24 17:31  dlc

	* MANIFEST (1.6): Added another test.

2002-10-24 17:31  dlc

	* TabularDisplay.pm (1.17): Added paginate() method.

2002-10-16 07:04  dlc

	* TabularDisplay.pm (1.16): populate unintentionally ate the array
	it was passed; fixed.  Added items method, which returns the number
	of elements that the structure contains.

2002-10-15 16:26  dlc

	* examples/mysql.pl (1.2): Added command history (reads/writes to
	~/.mysql_history).

2002-10-15 16:21  dlc

	* examples/aliases (1.2): Added -l to #! line.

2002-10-15 16:21  dlc

	* t/08.t (1.5): Removed alignment tests.

2002-10-15 16:20  dlc

	* TabularDisplay.pm (1.15): Silly diff: added ; to end of a line.

2002-10-15 16:20  dlc

	* TabularDisplay.pm (1.14): Moved Version() to beginning.  Removed
	references to _format_column, including definition.  Removed _align
	function: Alignment no longer exists.

2002-10-09 07:19  dlc

	* TabularDisplay.pm (1.12), TabularDisplay.pm (1.13): Gets version
	from $Revision: 1.1.1.1 $.

2002-10-09 07:18  dlc

	* Makefile.PL (1.4): Uses TabularDisplay.pm for version.

2002-10-09 07:14  dlc

	* TabularDisplay.pm (1.11): Added _format_column function, which
	formats the column names.  Added _align function, which justifies
	text Added sum function, which returns the sum of the elements in
	an array.  This is a nice function, I like it.	Added integer
	pragma, in a (potentially misguided) attempt to make division
	faster.  Updated documentation to reflect changes.

2002-10-09 07:11  dlc

	* MANIFEST (1.5): Added another example and another test.

2002-10-09 07:11  dlc

	* t/08.t (1.4): Readded test, with new contents.

2002-10-09 07:05  dlc

	* examples/dump-passwd.pl (1.1): Another example.

2002-10-08 16:35  dlc

	* t/08.t (1.3): Removed string overloading, so this test is not
	relevent.

2002-10-08 13:47  dlc

	* TabularDisplay.pm (1.10): Clears _COLUMNS when new columns are
	added.

2002-10-08 13:46  dlc

	* t/13.t (1.1): Test for column silliness, reported by "David N.
	Blank-Edelman" <dnb@ccs.neu.edu>.

2002-09-24 06:47  dlc

	* TabularDisplay.pm (1.9): Automatically sets $VERSION

2002-09-24 06:39  dlc

	* TabularDisplay.pm (1.8), Makefile.PL (1.3): Incremented version
	number for release.

2002-09-24 06:38  dlc

	* Changes (1.7), Changes (1.8), Changes (1.9): Automatically
	generated by cvs2cl.pl for dlc@tumbleweed

2002-09-24 06:38  dlc

	* .todo (1.2): Marked documentation complete.

2002-09-23 23:43  dlc

	* TabularDisplay.pm (1.7): More VERSION stuff.

2002-09-23 23:41  dlc

	* TabularDisplay.pm (1.6): Added VERSION stuff.

2002-09-23 23:39  dlc

	* Changes (1.6): Automatically generated by cvs2cl.pl for
	dlc@tumbleweed

2002-09-23 23:39  dlc

	* Makefile.PL (1.2): Updated $VERSION

2002-09-23 23:35  dlc

	* t/05.t (1.3): Changes to accomidate $start and $end additions to
	render().

2002-09-23 23:35  dlc

	* README (1.3): Updated README

2002-09-23 23:35  dlc

	* Changes (1.4), Changes (1.5): Automatically generated by
	cvs2cl.pl for dlc@tumbleweed

2002-09-23 23:34  dlc

	* TabularDisplay.pm (1.5): Added $start and $end to render();
	documented them.

2002-09-23 23:23  dlc

	* TabularDisplay.pm (1.4): Some more documentation.

2002-09-23 22:44  dlc

	* MANIFEST (1.4): Added examples/mysql.pl; resorted

2002-09-23 22:44  dlc

	* examples/mysql.pl (1.1): Example mysql client, using
	Text::TabularDisplay

2002-09-23 18:15  dlc

	* MANIFEST (1.3): Added examples and t/12.t

2002-09-23 18:14  dlc

	* README (1.2): Updated README.

2002-09-23 18:12  dlc

	* Changes (1.3): Automatically generated by cvs2cl.pl for
	dlc@tumbleweed

2002-09-23 18:12  dlc

	* TabularDisplay.pm (1.3): Moved resetting to reset() and out of
	columns().  columns() calls _add internally.  add() calls _add
	internally.  Added _add() function, which adds data to the
	instance.  Removed references to $self->{ _COLUMNS }; replaced with

examples/mysql.pl
#!/usr/bin/perl

# ----------------------------------------------------------------------
# $Id: mysql.pl,v 1.1.1.1 2005/10/04 16:39:58 dlc Exp $
# ----------------------------------------------------------------------
# Example usage of Text::TabularDisplay to implement a low-functionality
# version of the mysql text monitor.
# ----------------------------------------------------------------------

use strict;
use vars qw(%opts $prog $histfile);

use Carp qw(carp);
use DBI;
use File::Basename qw(basename);
use File::Spec;
use Getopt::Long;
use Term::ReadLine;
use Text::TabularDisplay;

$prog = basename $0;
$histfile = File::Spec->catfile($ENV{'HOME'}, ".mysql_history");

eval {
    require MySQL::Config;
    my %defaults = MySQL::Config::parse_defaults("my", [ qw(client) ]);

    for (qw(user password host)) {
        $opts{$_} = $defaults{$_};
    }
};

GetOptions(\%opts,
    "user|u=s",
    "password|p=s",
    "help|?!",
    "host=s");

$opts{'host'} ||= "localhost";
if (defined $opts{'help'}) {
    print STDERR "$prog - mysql command line client emulation\n",
                 "Usage: $prog [OPTIONS] DB_NAME\n\n",
                 "OPTIONS include:\n",
                 "  --user=\$username      Username to connect as\n",
                 "  --password=\$password  Password for \$user\n",
                 "  --host=\$host          Host on which DB_NAME can be found\n",
                 "  --help                You're reading it.\n\n";
    exit(1);
}


my $db = shift(@ARGV) or die "$prog: Must supply a database name!\n";

# Create the Text::TabularDisplay and Term::ReadLine instances, and
# make the database connection.
my $table = Text::TabularDisplay->new;
my $term = Term::ReadLine->new("mysql");
my $dbh = DBI->connect("dbi:mysql:database=$db;host=$opts{'host'}",
                        $opts{'user'}, $opts{'password'})
    or die "Can't connect to $db on $opts{'host'}: $DBI::errstr";

$term->ReadHistory($histfile);

while (defined (my $line = $term->readline("mysql> "))) {
    #$term->AddHistory($line);
    my $sth;

    if ($line =~ /^\s*(quit|exit)/) {
        last;
    }

    unless ($sth = $dbh->prepare($line)) {
        carp "Can't prepare line: " . $dbh->errstr;
        next;
    }

    # Reset the table
    $table->reset;

    unless ($sth->execute) {
        carp "Can't execute query: " . $sth->errstr;
        next;
    }

    # Set the columns
    my $names = $sth->{'NAME'};
    $table->columns(@$names);

    while (my $row = $sth->fetchrow_arrayref) {
        # Add data to the table
        $table->add($row);
    }
    $sth->finish;

    # Print the final version of the table
    # Note that without the trailing \n, the last line is buffered,
    # which is pretty ugly...
    printf "%s\n", $table->render;
}

$term->WriteHistory($histfile);

print "Bye!\n";
exit(0);
